### PR TITLE
Replace busy-wait atomic bool goto logic

### PR DESCRIPTION
**Pull request by kklobe**
This replaces the triangle worker for loop/goto/atomic bool busy-wait logic with an std::condition_variable and std::mutex which more cleanly expresses the intent.

It also improves performance (disclaimer: local microbenchmark on Apple Silicon M1):

main:
```
2023-05-27 17:14:07.819 | triangle_worker_run: 13 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 7 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 11 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 8 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 14 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 8 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 7 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 10 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 5 pixels per us
2023-05-27 17:14:07.819 | triangle_worker_run: 8 pixels per us
```

PR:
```
2023-05-27 17:15:27.843 | triangle_worker_run: 23 pixels per us
2023-05-27 17:15:27.843 | triangle_worker_run: 12 pixels per us
2023-05-27 17:15:27.843 | triangle_worker_run: 17 pixels per us
2023-05-27 17:15:27.843 | triangle_worker_run: 20 pixels per us
2023-05-27 17:15:27.843 | triangle_worker_run: 25 pixels per us
2023-05-27 17:15:27.844 | triangle_worker_run: 22 pixels per us
2023-05-27 17:15:27.844 | triangle_worker_run: 15 pixels per us
2023-05-27 17:15:27.844 | triangle_worker_run: 20 pixels per us
2023-05-27 17:15:27.844 | triangle_worker_run: 22 pixels per us
2023-05-27 17:15:27.844 | triangle_worker_run: 22 pixels per us
```

the benchmark snippet used in `triangle_worker_run()`:
```cpp
	const auto start = std::chrono::steady_clock::now();
	{
		std::lock_guard lock(tworker.done_mutex);
		tworker.done_count = 0;
	}
	for (size_t i = 0; i != TRIANGLE_THREADS; i++) {
		SDL_SemPost(tworker.sembegin[i]);
	}
	triangle_worker_work(tworker, TRIANGLE_THREADS, TRIANGLE_WORKERS);
	{
		std::unique_lock lock(tworker.done_mutex);
		tworker.done_cv.wait(lock, [&] {
			return tworker.done_count == TRIANGLE_THREADS;
		});
	}
	const auto stop = std::chrono::steady_clock::now();
	const auto duration = std::chrono::duration_cast<std::chrono::microseconds>(stop - start);
	const auto pix_per_us = tworker.totalpix / duration.count();
	LOG_INFO("triangle_worker_run: %lld pixels per us", pix_per_us);
```

### CODE REVIEW

**Comment by kcgen** [comment]
Will let reviewers assess this for 3 more days - then going ahead.

File: src/hardware/voodoo.cpp [hunk 1]
```diff
@@ -937,7 +938,9 @@ struct triangle_worker
 	poly_vertex v1, v2, v3;
 	INT32 v1y, v3y, totalpix;
 	SDL_sem* sembegin[TRIANGLE_THREADS];
-	std::atomic_bool done[TRIANGLE_THREADS];
+	std::condition_variable done_cv;
+	std::mutex done_mutex;
+	int done_count;
```
Comment (kcgen) [inline src/hardware/voodoo.cpp:943]: Suggest defaults for the struct members (might as well do them all).

**Review by kcgen — APPROVED** [review]
Ok, it's sat 5 days without feedback from the requested reviews, so I'll review.

Looks good, @kklobe.
Just a minor suggestion.

File: src/hardware/voodoo.cpp (hunk 1 above)
Comment (kklobe) [reply src/hardware/voodoo.cpp:943]: does the

```cpp
	triangle_worker tworker = {};
```

in `voodoo_state` handle that? And if not, should I then init all the members in triangle_worker, in addition to my new ones?

Comment (kcgen) [reply src/hardware/voodoo.cpp:943]: hmm.. I guess this starts to ripple further into more types, like `poly_vertex`.  

If it's passing CI then at least `-Weffc++` is satisfied there aren't any undefined states, so might as well defer my comment for a focused cleanup PR.

**Comment by kcgen** [comment]
Looks great, @kklobe!

**Pull request merged**